### PR TITLE
docs(blog): running python with WebAssembly part 1

### PR DESCRIPTION
**Pull request by michael-0acf4**
Running python with webassembly (part 1)

#### Migration notes

None

- [ ] The change comes with new or modified tests
- [ ] Hard-to-understand functions have explanatory comments
- [x] End-user documentation is updated to reflect the change

## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive guide for integrating Python runtime with WebAssembly (WASI) in the Metatype ecosystem.
  - Detailed the advantages of using WebAssembly over Docker for platform independence and resource management.
  - Provided technical requirements and a refined solution for executing Python scripts in a sandboxed environment.
  - Expanded vocabulary with new relevant terms for enhanced text processing and validation. 

- **Documentation**
  - Updated YAML configuration structure in documentation for clarity on type gate usage.

### CODE REVIEW

**Comment by michael-0acf4** [comment]
@coderabbitai review

**Review by zifeo — APPROVED** [review]
Maybe uppercasing the tech name? e.g. Metatype, Deno?

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+## The WebAssembly Interface (WASI)
+
+This is basically WebAssembly but with more capabilities such as file system access and networking.
+
+Just like x86, or Wasm, WASI is a compilation target.
+The **build** can then be executed in a sandboxed environment within a WASI-compliant runtime (WasmEdge, Wasmtime, Wasmer).
+
+This fits very well to our use case, and as opposed to Docker, we can instantiate, manage and expose ressources programmatically with ease.
+
+And luckily for us at the time, some [community works](https://github.com/vmware-labs/webassembly-language-runtimes) has already been done on enabling the use of Python into WASI.
```
Comment (Yohe-Am) [inline website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:45]: I don't think "work" should be plural here.

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+To complicate things a bit, the WASI spec itself is a work in progress.
+The one that is supported the most is the [preview1](https://github.com/WebAssembly/WASI/blob/main/legacy/preview1/docs.md).
+
+### File system access
+
+Since we relied on PyO3's abi3-py311 for running Python script in rust, it required libpython to be available at runtime.
```
Comment (Yohe-Am) [inline website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:52]: PyO3 is not properly introduced here. A sentence or two?

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+2. **Speed**: functions should run with low latency
+3. **User-friendliness**: writing a function in one language should be as near identical as writing one in another
+
+## Choosing an approach...
+
+Let's start simple, we want to run Python code, so what do we need?
+
+- A Python runtime
+
+We do not want to be platform-dependent, but how can we achieve that?
```
Comment (Yohe-Am) [inline website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:29]: A few more words towards the limitation of the python runtime might be good. We can do this on Deno but not python, why?

**Review by Yohe-Am — COMMENTED** [review]
I like what's there so far but I have some comments. Willing to address them myself if you don't mind. Might have to rework the structure to better introduce the second section.

**Comment by michael-0acf4** [comment]
> I like what's there so far but I have some comments. Willing to address them myself if you don't mind. Might have to rework the structure to better introduce the second section.

@Yohe-Am feel free to do so, I sort of wrote it in a way that it can be merged as is.

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx (hunk 3 above)
Comment (michael-0acf4) [reply website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:29]: Good point. We can provide an overview of how we did it with Deno as a parallel.

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx (hunk 1 above)
Comment (michael-0acf4) [reply website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:45]: You're right, it's surprising that CodeRabbit didn't catch that!

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+- Docker
+- WebAssembly
+
+At a higher level, Docker and WebAssembly all share the same idea: package an application into one build and make it platform independent.
+
+## The WebAssembly Interface (WASI)
```
Comment (Natoandro) [inline website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:36]: "System" Interface?

**Review by Natoandro — APPROVED** [review]

File: website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+This approach turned out to be quite hacky.
+
+After encountering several issues, we ultimately decided to use **preopens**, which provided the features we needed for our situation.
+
+```rust
+fn init_Python_vm() -> Result<Vm> {
+  let preopens = vec![
+    // User script will be uploaded at ./src/Python which is virtually seen as /app
+    // Each script only has readonly access to /app
+    "/app:./src/Python:readonly".to_owned()
```
Comment (michael-0acf4) [inline website/blog/2024-08-26-how-we-run-python-using-webassembly/index.mdx:73]: "only" is redundant

**Comment by Yohe-Am** [comment]
@coderabbitai review

File: docs/metatype.dev/blog/2024-08-26-python-on-webassembly/index.mdx [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+  // ..
+
+  Ok(rt)
+}
+```
+
+### WASI 0.2
+
+The solution described above worked well to an extent (find the code [here](https://github.com/metatypedev/python-wasi-reactor)) but the limitations of preview1 and all the wrangling with PyO3 resulted in complexity that we were always ready to get rid of.
```
[the remark this answers is not in the document]
Comment (zifeo) [reply docs/metatype.dev/blog/2024-08-26-python-on-webassembly/index.mdx:113]: @michael-0acf4 totally, but we can reopen it if relevant?

**Review by destifo — APPROVED** [review]
Lgtm! there's one image that doesn't load on the hackmd version of the blog, I'm guessing it's on the hackmd side only

**Review by Yohe-Am — APPROVED** [review]

**Review by Natoandro — APPROVED** [review]

**Pull request merged**